### PR TITLE
Custom Actions: call outs to external APIs

### PR DESCRIPTION
**Pull request by snopoke**
## Description
This adds a 'custom actions' feature to OCS. A "custom action" represents an external API. When a bot is configured with a custom action the bot will have additional tools which will allow it to make HTTP requests to the external API.

## User Impact
Users can configure bots with custom actions.

### Demo
<div>
    <a href="https://www.loom.com/share/1deaa9eeb37d4367962e793a3a406be6">
      <p>Custom Actions Demo 🤖 - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/1deaa9eeb37d4367962e793a3a406be6">
      
    </a>
  </div>

### Docs
TODO

### CODE REVIEW

**Comment by snopoke** [comment]
@coderabbit review

**Comment by snopoke** [comment]
@coderabbit review

File: apps/chat/agent/openapi_tool.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.executors = [ActionExecutor(action) for action in self.custom_actions]
+
+    def _run(self, endpoint: str, method: str, params: dict | None, data: dict | None, headers: dict | None) -> str:
+        for executor in self.executors:
+            if executor.can_run(endpoint, method):
+                return executor.run(endpoint, method, params, data, headers)
```
Comment (SmittieC) [inline apps/chat/agent/openapi_tool.py:53]: Does the fact that we're returning here mean that only one executer will ever be run (assuming all can actuall run). What if I have multiple `custom_actions` configured?

File: apps/custom_actions/fields.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import json
+
+import yaml
+from django.core.exceptions import ValidationError
+from django.forms import Textarea
+from django.forms.fields import CharField, InvalidJSONInput, JSONString
+
+
+class JSONOrYAMLField(CharField):
```
Comment (SmittieC) [inline apps/custom_actions/fields.py:9]: `JsonOrYamlField`?

File: apps/chat/agent/openapi_schema.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+class FunctionDef(TypedDict):
+    name: str
+    description: str
+    method: str
+    url: str
+    args_schema: type[BaseModel]
+
+
+def openapi_spec_op_to_function_def(spec: OpenAPISpec, path: str, method: str) -> FunctionDef:
```
Comment (SmittieC) [inline apps/chat/agent/openapi_schema.py:18]: Where is this function being used?

File: apps/chat/agent/openapi_tool.py (hunk 1 above)
Comment (SmittieC) [reply apps/chat/agent/openapi_tool.py:53]: Oh I see now, the LLM will specify the endpoint + method to run. We then loop through all schemata to check if the endpoint exists in it. If it does, we mark that one as runnable. So we basically hit the first `endpoint` with `method` that exists across all specified `custom_actions`.

**Review by SmittieC — COMMENTED** [review]
Nice

File: apps/chat/agent/openapi_schema.py (hunk 3 above)
Comment (snopoke) [reply apps/chat/agent/openapi_schema.py:18]: It isn't yet - I'm going to refactor the Tool now to use this instead of the current method.

**Comment by snopoke** [comment]
@coderabbit review

**Comment by snopoke** [comment]
@coderabbit resolve

File: apps/utils/urlvalidate.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            ip_address.is_multicast,
+            ip_address.is_private,
+        ]
+    ):
+        raise PossibleSSRFAttempt(f"Unsafe IP address: {ip_address}")
+    else:
+        return ip_address
+
+
+class PossibleSSRFAttempt(Exception):
```
Comment (SmittieC) [inline apps/utils/urlvalidate.py:69]: SSRF or CSRF?

File: apps/utils/urlvalidate.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+# Copied from CommCare HQ: https://github.com/dimagi/commcare-hq/blob/master/corehq/util/urlvalidate/urlvalidate.py#L6
```
Comment (SmittieC) [inline apps/utils/urlvalidate.py:1]: :+1:

**Review by SmittieC — DISMISSED** [review]
I did not go through it in super detail, but overall it looks great. I like the fact that you're checking the url for a possible attack.

**Pull request merged**